// vike/dist/cjs/node/plugin/plugins/importUserCode/v1-design/getVikeConfig/resolveFilePath.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveFilePathAbsoluteFilesystem = exports.resolveFilePathRelativeToUserRootDir = void 0;
const path_1 = __importDefault(require("path"));
const utils_js_1 = require("../../../../utils.js");
function resolveFilePathRelativeToUserRootDir(filePathRelativeToUserRootDir, userRootDir) {
    (0, utils_js_1.assertPosixPath)(filePathRelativeToUserRootDir);
    (0, utils_js_1.assertPosixPath)(userRootDir);
    const filePathAbsoluteFilesystem = path_1.default.posix.join(userRootDir, filePathRelativeToUserRootDir);
    return getFilePathResolved(filePathRelativeToUserRootDir, filePathAbsoluteFilesystem);
}
exports.resolveFilePathRelativeToUserRootDir = resolveFilePathRelativeToUserRootDir;
function resolveFilePathAbsoluteFilesystem(filePathAbsoluteFilesystem, userRootDir) {
    (0, utils_js_1.assertPosixPath)(filePathAbsoluteFilesystem);
    (0, utils_js_1.assertPosixPath)(userRootDir);
    let filePathRelativeToUserRootDir = path_1.default.posix.relative(userRootDir, filePathAbsoluteFilesystem);
    (0, utils_js_1.assert)(!filePathRelativeToUserRootDir.startsWith('.') && !filePathRelativeToUserRootDir.startsWith('/'));
    filePathRelativeToUserRootDir = '/' + filePathRelativeToUserRootDir;
    return getFilePathResolved(filePathRelativeToUserRootDir, filePathAbsoluteFilesystem);
}
exports.resolveFilePathAbsoluteFilesystem = resolveFilePathAbsoluteFilesystem;
function getFilePathResolved(filePathRelativeToUserRootDir, filePathAbsoluteFilesystem) {
    return {
        filePathRelativeToUserRootDir,
        filePathAbsoluteVite: filePathRelativeToUserRootDir,
        filePathAbsoluteFilesystem,
        filePathToShowToUser: filePathRelativeToUserRootDir,
        importPathAbsolute: null
    };
}
