// vike/dist/cjs/utils/nodeEnv.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.assertNodeEnvIsNotDev = exports.getNodeEnvDesc = exports.isNodeEnvDev = exports.setNodeEnvToProduction = exports.getNodeEnv = void 0;
const picocolors_1 = __importDefault(require("@brillout/picocolors"));
const assertIsNotBrowser_js_1 = require("./assertIsNotBrowser.js");
const assert_js_1 = require("./assert.js");
(0, assertIsNotBrowser_js_1.assertIsNotBrowser)();
function getNodeEnv() {
    if (typeof process === 'undefined')
        return null;
    return process.env.NODE_ENV;
}
exports.getNodeEnv = getNodeEnv;
function setNodeEnvToProduction() {
    // The statement `process.env['NODE_ENV'] = 'production'` chokes webpack v4
    const proc = process;
    const { env } = proc;
    env.NODE_ENV = 'production';
}
exports.setNodeEnvToProduction = setNodeEnvToProduction;
function isNodeEnvDev() {
    const nodeEnv = getNodeEnv();
    if (!nodeEnv)
        return true;
    if (['development', 'dev'].includes(nodeEnv))
        return true;
    // That's quite aggressive, let's see if some user complains
    return false;
}
exports.isNodeEnvDev = isNodeEnvDev;
function getNodeEnvDesc() {
    const nodeEnv = getNodeEnv();
    const isDev = isNodeEnvDev();
    const nodeEnvDesc = `environment is set to be a ${(isDev ? 'development' : 'production')} environment by ${picocolors_1.default.cyan(`process.env.NODE_ENV === ${JSON.stringify(nodeEnv)}`)}`;
    return nodeEnvDesc;
}
exports.getNodeEnvDesc = getNodeEnvDesc;
function assertNodeEnvIsNotDev(operation) {
    const isDev = isNodeEnvDev();
    if (!isDev)
        return;
    const nodeEnvDesc = getNodeEnvDesc();
    (0, assert_js_1.assertUsage)(false, `The ${nodeEnvDesc} which is forbidden upon ${operation}, see https://vike.dev/NODE_ENV`);
}
exports.assertNodeEnvIsNotDev = assertNodeEnvIsNotDev;
