// vike/dist/esm/node/plugin/plugins/importUserCode/v1-design/getVikeConfig/resolveFilePath.js
export { resolveFilePathRelativeToUserRootDir };
export { resolveFilePathAbsoluteFilesystem };
import path from 'path';
import { assert, assertPosixPath } from '../../../../utils.js';
function resolveFilePathRelativeToUserRootDir(filePathRelativeToUserRootDir, userRootDir) {
    assertPosixPath(filePathRelativeToUserRootDir);
    assertPosixPath(userRootDir);
    const filePathAbsoluteFilesystem = path.posix.join(userRootDir, filePathRelativeToUserRootDir);
    return getFilePathResolved(filePathRelativeToUserRootDir, filePathAbsoluteFilesystem);
}
function resolveFilePathAbsoluteFilesystem(filePathAbsoluteFilesystem, userRootDir) {
    assertPosixPath(filePathAbsoluteFilesystem);
    assertPosixPath(userRootDir);
    let filePathRelativeToUserRootDir = path.posix.relative(userRootDir, filePathAbsoluteFilesystem);
    assert(!filePathRelativeToUserRootDir.startsWith('.') && !filePathRelativeToUserRootDir.startsWith('/'));
    filePathRelativeToUserRootDir = '/' + filePathRelativeToUserRootDir;
    return getFilePathResolved(filePathRelativeToUserRootDir, filePathAbsoluteFilesystem);
}
function getFilePathResolved(filePathRelativeToUserRootDir, filePathAbsoluteFilesystem) {
    return {
        filePathRelativeToUserRootDir,
        filePathAbsoluteVite: filePathRelativeToUserRootDir,
        filePathAbsoluteFilesystem,
        filePathToShowToUser: filePathRelativeToUserRootDir,
        importPathAbsolute: null
    };
}
